crates/runtime-table-partition/src/provider.rs

```diff
@@ -111,13 +141,148 @@ impl TableProvider for PartitionTableProvider {
         TableType::Base
     }
 
+    fn supports_filters_pushdown(
+        &self,
+        filters: &[&Expr],
+    ) -> Result<Vec<TableProviderFilterPushDown>, DataFusionError> {
+        Ok(vec![TableProviderFilterPushDown::Exact; filters.len()])
+    }
+
     async fn scan(
+        &self,
+        state: &dyn Session,
+        projection: Option<&Vec<usize>>,
+        filters: &[Expr],
+        limit: Option<usize>,
+    ) -> Result<Arc<dyn ExecutionPlan>, DataFusionError> {
+        let partitions = self.partitions.read().await;
+        let mut plans = Vec::with_capacity(partitions.len());
+        for partition in partitions.values() {
+            if prune_partition(filters, &self.partition_by, &partition.partition_value) {
+                continue;
+            }
+            let plan = partition
+                .table_provider
+                .scan(state, projection, filters, limit)
+                .await?;
+            plans.push(plan);
+        }
+
+        let plan = match plans {
+            plans if plans.is_empty() => {
+                return Ok(Arc::new(EmptyExec::new(Arc::clone(&self.schema))));
+            }
+            mut plans if plans.len() == 1 => plans.pop().ok_or_else(|| {
+                DataFusionError::Execution("expected an ExecutionPlan".to_string())
+            })?,
+            plans => Arc::new(UnionExec::new(plans)),
+        };
+
+        if let Some(limit) = limit {
+            return Ok(Arc::new(GlobalLimitExec::new(plan, limit, None)));
+        }
+
+        Ok(plan)
+    }
+
+    async fn insert_into(
         &self,
         _state: &dyn Session,
-        _projection: Option<&Vec<usize>>,
-        _filters: &[Expr],
-        _limit: Option<usize>,
+        input: Arc<dyn ExecutionPlan>,
+        insert_op: InsertOp,
     ) -> Result<Arc<dyn ExecutionPlan>, DataFusionError> {
-        todo!()
+        Ok(Arc::new(PartitionInsertExec::new(
+            input,
+            self.partition_by.clone(),
+            Arc::clone(&self.creator),
+            Arc::clone(&self.partitions),
+            insert_op,
+            Arc::clone(&self.schema),
+        )))
+    }
+}
+
+/// Determine whether a partition should be pruned from the scan plan based on
+/// the query `filters`, the expression that the partition was created from,
+/// `partition_by`, and the `partition_value` produced by the `partition_by`
+/// `Expr` for this particular partition.
+fn prune_partition(filters: &[Expr], partition_by: &Expr, partition_value: &ScalarValue) -> bool {
+    for filter in filters {
```

Comment:
Is there a card I missed that is tracking expanding this support to cover all expressions types?

Reply:
Also this is assuming that the partition by expression is a literal column - this should actually compute the expression in terms of the query filter.

Reply:
I've added an issue to address making this more generic https://github.com/spiceai/spiceai/issues/6406.